### PR TITLE
wifi: eswifi: Implement TCP/UDP socket offload with TLS

### PR DESCRIPTION
**Pull request by ssekar15**
Newly implement socket based communication on eswifi mainly to
achive TLS. Tested with Inventek ISM43362-M3G-L44.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

### CODE REVIEW

**Comment by ssekar15** [comment]
@erwango @loicpoulain With TLS based socket communication, data exchange between server(ubuntu) and client(Zephyr) happens once successfully. 2nd data is sent from server and ACK by eswifi TCP stack but data is not received at eswifi driver. Same scenario apply to raw TLS socket application and mqtt examples.

Note: Socket communication without TLS works

Screenshot from 2019-08-23 22-45-49

**Comment by loicpoulain** [comment]
Is socket offload supposed to fully replace net_context offload at some point?

**Comment by jukkar** [comment]
> Is socket offload supposed to fully replace net_context offload at some point?

They are different level offloading, the other one is at BSD socket level, and the other one is even lower level. There has not been discussion whether we should remove net_context offloading.

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+	.connect = eswifi_socket_connect,
+	.setsockopt = eswifi_socket_setsockopt,
+	.recv = eswifi_socket_recv,
+	.send = eswifi_socket_send,
+	.poll = eswifi_socket_poll,
+};
+
+int eswifi_socket_offload_init(struct eswifi_dev *leswifi)
+{
+#ifdef CONFIG_NET_SOCKETS_OFFLOAD
```
Comment (loicpoulain) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: Why? eswifi_socket_offload is not built anyway if CONFIG_NET_SOCKETS_OFFLOAD is not set

**Comment by loicpoulain** [comment]
@ssekar15, Overall it looks good, but a lot of code is duplicate of net_context offload (eswifi_offload) and so I would prefer to have this factorized in a shared place, which should simplify maintenance. Is it something you can do?

For TLS issue, which does not seem to be a blocker for merge, we'll had to create an issue to track the problem and hopefully fix it later.

**Review by loicpoulain — CHANGES_REQUESTED** [review]

**Comment by ssekar15** [comment]
> factorized in a shared place

Yes, I will do.

File: drivers/wifi/eswifi/eswifi_socket_offload.c (hunk 1 above)
Comment (ssekar15) [reply drivers/wifi/eswifi/eswifi_socket_offload.c]: It was from draft version, I will remove it

**Review by vanti — CHANGES_REQUESTED** [review]
A few minor comments - looks fine to me otherwise.

**Comment by loicpoulain** [comment]
@ssekar15, thanks looks better, you just need to address the few minors comments, also I would rename 'eswifi_common.c' to 'eswifi_socket.c' since all functions are related to socket management (and other generic functions are in eswifi_core).

**Comment by ssekar15** [comment]
> @ssekar15, thanks looks better, you just need to address the few minors comments, also I would rename 'eswifi_common.c' to 'eswifi_socket.c' since all functions are related to socket management (and other generic functions are in eswifi_core).

@loicpoulain Done, please review

**Review by vanti — APPROVED** [review]

File: drivers/wifi/eswifi/eswifi_offload.c [hunk 2]
```diff
@@ -575,31 +385,21 @@ static int eswifi_off_put(struct net_context *context)
...
-	}
-
-	__select_socket(eswifi, socket->index);
-
-	k_delayed_work_cancel(&socket->read_work);
-
-	__select_socket(eswifi, socket->index);
-	__stop_socket(eswifi, socket);
+	ret = __eswifi_socket_free(eswifi, socket);
+	if (ret)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_offload.c]: Style issue, one liners need to have {}

File: drivers/wifi/eswifi/eswifi_offload.c [hunk 3]
```diff
@@ -575,31 +385,21 @@ static int eswifi_off_put(struct net_context *context)
...
+		goto done;
 
-	if (--socket->usage <= 0) {
+	if (--socket->usage <= 0)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_offload.c]: same here

File: drivers/wifi/eswifi/eswifi_offload.c [hunk 4]
```diff
@@ -609,75 +409,32 @@ static int eswifi_off_get(sa_family_t family,
...
-	} else {
-		LOG_ERR("Only TCP & UDP is supported");
-		eswifi_unlock(eswifi);
-		return -EPFNOSUPPORT;
-	}
-
-	snprintf(eswifi->buf, sizeof(eswifi->buf), "P1=%d\r", socket->type);
+	idx = __eswifi_socket_new(eswifi, family, type, ip_proto, *context);
+	if (idx < 0)
+		goto unlock;
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_offload.c]: here too

File: drivers/wifi/eswifi/eswifi_socket.c [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "eswifi.h"
+#include <net/net_pkt.h>
+
+#define LOG_LEVEL CONFIG_WIFI_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wifi_eswifi_common);
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket.c]: You can also write this into one line like this so the LOG_LEVEL do not need to be defined.
```
LOG_MODULE_REGISTER(wifi_eswifi, CONFIG_WIFI_LOG_LEVEL);
```
Also the module can be set to `wifi_eswifi`

Please put the logging include before the zephyr.h. That way if some of the header files have inline functions with logging, the output is printed properly.

File: drivers/wifi/eswifi/eswifi_socket.c [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+int __eswifi_socket_new(struct eswifi_dev *eswifi, int family, int type,
+			int proto, void *context)
+{
+	struct eswifi_off_socket *socket = NULL;
+	int err, i;
+
+	LOG_DBG("");
+
+	if (family != AF_INET) {
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket.c]: BTW, do you have somewhere a check whether IPv4 is enabled at all?

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 7]
```diff
@@ -0,0 +1,391 @@
+/*
+ * Copyright (c) 2019 Linumiz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_LEVEL CONFIG_WIFI_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wifi_eswifi_socket_offload);
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: Just do `LOG_MODULE_REGISTER(wifi_eswifi, CONFIG_WIFI_LOG_LEVEL);` and remove the LOG_LEVEL define.

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+}
+
+static int eswifi_socket_accept(int sock, struct sockaddr *addr,
+				socklen_t *addrlen)
+{
+	struct eswifi_off_socket *socket;
+	int ret;
+
+	if ((addrlen == NULL) || (addr == NULL) ||
+	    (sock > ESWIFI_OFFLOAD_MAX_SOCKETS))
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: Even one liners need {}

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+{
+	sec_tag_t *sec_tags = (sec_tag_t *)optval;
+	int ret = 0;
+	int tags_len;
+	sec_tag_t tag;
+	int id;
+	int i, bytes;
+	struct tls_credential *cert;
+
+	if ((optlen % sizeof(sec_tag_t)) != 0 || (optlen == 0))
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: same here

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+				id = 2;
+				break;
+			case TLS_CREDENTIAL_NONE:
+			case TLS_CREDENTIAL_PSK:
+			case TLS_CREDENTIAL_PSK_ID:
+			default:
+				/* Not handled */
+				return -EINVAL;
+			}
+			snprintf(eswifi->buf, sizeof(eswifi->buf),
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: empty line after }

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+				 "PG=%d,%d,%d\r", 0, id, cert->len);
+			bytes = strlen(eswifi->buf);
+			memcpy(&eswifi->buf[bytes], cert->buf, cert->len);
+			bytes += cert->len;
+			LOG_DBG("cert write len %d\n", cert->len);
+			ret = eswifi_request(eswifi, eswifi->buf, bytes + 1,
+				     eswifi->buf, sizeof(eswifi->buf));
+			LOG_DBG("cert write err %d\n", ret);
+			if (ret < 0)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+				return ret;
+
+			snprintf(eswifi->buf, sizeof(eswifi->buf), "PF=0,0\r");
+			ret = eswifi_at_cmd(eswifi, eswifi->buf);
+			if (ret < 0)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: ditto

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+}
+
+static ssize_t eswifi_socket_send(int sock, const void *buf, size_t len,
+				  int flags)
+{
+	struct eswifi_off_socket *socket;
+	int ret;
+	int offset;
+
+	if (!buf)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: ditto

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+}
+
+static ssize_t eswifi_socket_recv(int sock, void *buf, size_t max_len,
+				  int flags)
+{
+	struct eswifi_off_socket *socket;
+	int len = 0, ret = 0;
+	struct net_pkt *pkt;
+
+	if ((max_len == 0) || (buf == NULL) ||
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: ditto

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+		return -EINVAL;
+
+	eswifi_lock(eswifi);
+	socket = &eswifi->socket[sock];
+
+	if (socket->prev_pkt_rem) {
+		pkt = socket->prev_pkt_rem;
+		goto skip_wait;
+	}
+	pkt = k_fifo_get(&socket->fifo, K_NO_WAIT);
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: empty line after }

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+
+skip_wait:
+	len = net_pkt_remaining_data(pkt);
+	if (len > max_len) {
+		len = max_len;
+		socket->prev_pkt_rem = pkt;
+	} else {
+		socket->prev_pkt_rem = NULL;
+	}
+	ret = net_pkt_read(pkt, buf, len);
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: same here

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+
+	if (!socket->prev_pkt_rem)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+		net_pkt_unref(pkt);
+done:
+	LOG_DBG("read %d %d %p", len, ret, pkt);
+	eswifi_unlock(eswifi);
+	if (ret)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+	return len;
+}
+
+static int eswifi_socket_close(int sock)
+{
+	struct eswifi_off_socket *socket;
+	struct net_pkt *pkt;
+	int ret;
+
+	if (sock > ESWIFI_OFFLOAD_MAX_SOCKETS)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+		return -EINVAL;
+
+	eswifi_lock(eswifi);
+
+	socket = &eswifi->socket[sock];
+	ret = __eswifi_socket_free(eswifi, socket);
+	if (ret)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 21]
```diff
@@ -0,0 +1,391 @@
...
+		goto done;
+
+	/* consume all net pkt */
+	while (1) {
+		pkt = k_fifo_get(&socket->fifo, K_NO_WAIT);
+		if (!pkt)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 22]
```diff
@@ -0,0 +1,391 @@
...
+static int eswifi_socket_open(int family, int type, int proto)
+{
+	struct eswifi_off_socket *socket = NULL;
+	int idx;
+
+	eswifi_lock(eswifi);
+
+	/* Assign dummy context SOCkEt(50CE) */
+	idx = __eswifi_socket_new(eswifi, family, type, proto, 0x50CE);
+	if (idx < 0)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 23]
```diff
@@ -0,0 +1,391 @@
...
+	eswifi_unlock(eswifi);
+	return idx;
+}
+
+static int eswifi_socket_poll(struct pollfd *fds, int nfds, int msecs)
+{
+	struct eswifi_off_socket *socket;
+	int sock, ret;
+
+	if (nfds > 1 ||
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 24]
```diff
@@ -0,0 +1,391 @@
...
+	    fds[0].fd  > ESWIFI_OFFLOAD_MAX_SOCKETS)
+		return -EINVAL;
+
+	eswifi_lock(eswifi);
+	sock = fds[0].fd;
+	socket = &eswifi->socket[sock];
+	eswifi_unlock(eswifi);
+	if (socket->state != ESWIFI_SOCKET_STATE_CONNECTED)
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

File: drivers/wifi/eswifi/eswifi_socket_offload.c [hunk 25]
```diff
@@ -0,0 +1,391 @@
...
+	return ret;
+}
+
+static int ewifi_socket_bind(int sock, const struct sockaddr *addr,
+			     socklen_t addrlen)
+{
+	struct eswifi_off_socket *socket;
+	int ret;
+
+	if ((addrlen == NULL) || (addr == NULL) ||
```
Comment (jukkar) [inline drivers/wifi/eswifi/eswifi_socket_offload.c]: braces {} missing

**Review by jukkar — CHANGES_REQUESTED** [review]

File: drivers/wifi/eswifi/eswifi_socket.c (hunk 6 above)
Comment (ssekar15) [reply drivers/wifi/eswifi/eswifi_socket.c]: Yes it is handled in eswifi_core.c

**Comment by jukkar** [comment]
@ssekar15 There was some comments that you missed, can you check those too?

**Comment by ssekar15** [comment]
> @ssekar15 There was some comments that you missed, can you check those too?

@jukkar Done, please review

**Review by jukkar — APPROVED** [review]

**Review by loicpoulain — APPROVED** [review]
LGTM

**Pull request merged**